Dynamic entity creation for tplink

rytilahti opened the pull request:
### This PR was part of the tplink modernizing effort, and is now obsolete as it's been split up into separate PRs:
- #117758
- #117759
- #117760
- #117761
- #117762
- #117764
- #117765
- #116605

Previous description
------

* Use upstream-provided information for creating available entities, allowing exposing more features and moves the maintenance burden from homeassistant to the upstream library.
* Add new platforms: binary_sensor, button, number, select, climate, update
* Expose sub-devices as their own devices that define `via_device`: https://github.com/home-assistant/architecture/discussions/1036

Testing requires master branch from python-kasa, https://github.com/python-kasa/python-kasa/issues/783 lists known issues and/or changes TBD.

![image](https://github.com/python-kasa/python-kasa/assets/3705853/b5b114db-835a-4d6e-a8df-8189e7d01f3e)


<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #111238
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

rytilahti committed:
Rename descriptor to feature

rytilahti committed:
Add binary_sensor

rytilahti committed:
Remove show_in_hass and category setting

rytilahti committed:
Don't try to add energy sensors on devices that do not support them

rytilahti committed:
Create always sensors for the parent device, as well as for all children no matter if the device is a strip

rytilahti committed:
Use via_device for child devices to let them have their own devices

rytilahti committed:
Initialize child coordinators only for smartstrip for the time being

bdraco commented on homeassistant/components/tplink/switch.py:
```diff
@@ -30,128 +30,79 @@ async def async_setup_entry(
+        self._attr_is_on = is_on
+
+        icon = self._feature.icon
+        # TODO: hacky way to support on/off icons. maybe this should be removed?
```
I think we can use icon translations instead?

rytilahti replied:
Indeed, we could definitely do that. I suppose we cannot provide these progmatically, but they need to be defined in the JSON file?

bdraco replied:
Right. They allow a default icon and one based on the current state

Example: https://github.com/home-assistant/core/blob/dev/homeassistant/components/tessie/icons.json

rytilahti replied:
Yes, I think we don't want to do that until we find a workable solution for this. I tried to look into how we could check for the existence of translation, but couldn't find a quick solution. #103294 is the relevant PR, adding just here to have it written down somewhere.

Given it is just UI candy, we can do without it for the time being.

rytilahti committed:
Use parent coordinator for binary sensors

rytilahti committed:
Add update platform

rytilahti committed:
Consolidate _handle_coordinator_update handling

rytilahti committed:
Add number platform

rytilahti committed:
Add some icon translations for testing

In order to work, this requires commenting out the icon definitions inside python-kasa

bdraco commented on homeassistant/components/tplink/number.py:
```diff
@@ -0,0 +1,91 @@
+"""Support for TPLink number entities."""
```
We can strip this out before merging and put it in a second PR

rytilahti replied:
Yes, this PR is just for development use, the final one will be split up, and the new platforms moved into separate PRs.

bdraco commented on homeassistant/components/tplink/update.py:
```diff
@@ -0,0 +1,105 @@
+"""Firmware update platform for tplink."""
```
Same for this one

rytilahti committed:
Handle unexpected exceptions when trying to access a feature value

rytilahti committed:
Don't update values during init

rytilahti committed:
Disable update platform for now

rytilahti committed:
Use upstream categories for entitycategory & generalize entity handling a bit

rytilahti committed:
Fix unique_id generation and disable debug entities per default

rytilahti committed:
Generalize entity creation

Not really happy with this yet, but at least the creation is now done in a single place that's easier to modify as needed

rytilahti committed:
Use device name as name for the primary switch

rytilahti committed:
Use unit for sensor

rytilahti committed:
Add initial climate platform

rytilahti committed:
use hasscompat from upstream, don't hide debug entities per default

rytilahti committed:
Add button platform

bdraco commented on homeassistant/components/tplink/binary_sensor.py:
```diff
@@ -0,0 +1,65 @@
+            translation_key=feature.id,
+            name=feature.name,
+            icon=feature.icon,
+            **feature.hass_compat.dict(),
```
Its a bit strange for the library to know about HomeAssistant.

I think we should rename this `legacy_compat` instead

rytilahti replied:
So this is a counterpart for https://github.com/python-kasa/python-kasa/pull/879 - basically, we need some way to inform homeassistant on its own special handling for some features. An alternative would be to store this information inside the integration, but I was thinking that it would be better to keep this local.

We could inverse it by 1) adding static ids inside python-kasa and 2) using those ids to map for wanted parameters. Do you think that would be a better approach?

bdraco replied:
We definitely shouldn't be teaching the library about Home Assistant.  We should store this information in the integration.

I think its fine to add information to `python-kasa` to make it easier to map in Home Assistant, but its separation of concerns problem for the library to start learning about HA

rytilahti replied:
https://github.com/python-kasa/python-kasa/pull/904 requires static identifiers, so we can move the mapping inside the integration.

bdraco commented on homeassistant/components/tplink/climate.py:
```diff
@@ -0,0 +1,142 @@
+    entities: list[TPLinkClimate] = []
+    # As there are no standalone thermostats, we just iterate over the children.
+    entities.extend(
+        [
+            TPLinkClimate(child, parent_coordinator)
+            for child in device.children
+            if child.device_type == DeviceType.Thermostat
+        ]
```
```suggestion
            TPLinkClimate(child, parent_coordinator)
            for child in device.children
            if child.device_type == DeviceType.Thermostat
```

rytilahti replied:
done.

rytilahti committed:
Add select platform

rytilahti committed:
Update binary_sensor attributes before adding it

rytilahti committed:
Use generator expression for constructing climate entities

sdb9696 reviewed (commented):
I think in general this is good to start moving forwards with breaking into smaller PRs.  At least if we can finalize the entity changes and one of the platforms, sensor or switch probably, we can start moving forwards in parallel.  Switch might be better as we can do that sooner while we iron out the extra HA info mappings.

@bdraco do you think it's a possibility to have a tplink development branch so we can start merging smaller PRs against an alpha release of python-kasa?

rytilahti committed:
Add a generic function to create *description, remove hass_extras

rytilahti committed:
Use upstream precision_hint for sensors

rytilahti commented on homeassistant/components/tplink/entity.py:
```diff
@@ -65,24 +75,181 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+            case Feature.Category.Primary:  # Main controls have no category
+                return None
+            case Feature.Category.Info:
+                return None
```
Changed in be3b586ce8d7

rytilahti committed:
Convert SmartDeviceException->KasaException, SmartDevice->Device

rytilahti committed:
Make mypy happier

rytilahti commented on homeassistant/components/tplink/sensor.py:
```diff
@@ -130,40 +138,91 @@ async def async_setup_entry(
     async_add_entities(entities)
 
 
+class Sensor(CoordinatedTPLinkEntity, SensorEntity):
```
Yeah, I kind of like shorter names as they are already namespaced by the package, but this probably needs to be changed to get it merged :-)

rytilahti committed:
Bump version to 0.7.0.dev0

rytilahti committed:
Fix linting

rytilahti commented:
All but the update platform has now their own PR, and the update will follow at some point when https://github.com/python-kasa/python-kasa/pull/856 gets done, so I think we can close this PR now.

@sdb9696, I'll go through your comments next time I have some more free cycles, but feel also free to create PRs on top of those PRs/branches below if you wish:

- #117758
- #117759
- #117760
- #117761
- #117762
- #117764
- #117765

The switch PR needs to be the first one to go as rest of them are based on it. One major missing piece (besides tests) I'm seeing is a way to map `(feature-id, *description extras)` to have proper icons, correct graphs for sensors etc.
